flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/catalog/CatalogCalciteSchema.java

```diff
@@ -0,0 +1,248 @@
+	public Set<String> getFunctionNames() {
+		return new HashSet<>();
+	}
+
+	@Override
+	public Expression getExpression(SchemaPlus parentSchema, String name) {
+		return  Schemas.subSchemaExpression(parentSchema, name, getClass());
+	}
+
+	@Override
+	public boolean isMutable() {
+		return true;
+	}
+
+	@Override
+	public Schema snapshot(SchemaVersion schemaVersion) {
+		return this;
+	}
+
+	/**
+	 * Register a ReadableCatalog to Calcite schema with given name.
+	 *
+	 * @param parentSchema	the parent schema
+	 * @param catalogName	name of the catalog
+	 * @param catalog	the catalog
+	 */
+	public static void registerCatalog(
+			SchemaPlus parentSchema,
+			String catalogName,
+			ReadableCatalog catalog) {
+		CatalogCalciteSchema newCatalog = new CatalogCalciteSchema(catalogName, catalog);
+		parentSchema.add(catalogName, newCatalog);
+
+		LOGGER.info("Registered catalog '{}' to Calcite", catalogName);
+	}
+
+	/**
+	 * A mapping between Flink Catalog's database and Calcite's schema.
+	 * Tables are registered as tables in the schema.
+	 */
+	private class DatabaseCalciteSchema implements Schema {
+
+		private final String dbName;
+		private final ReadableCatalog catalog;
+
+		public DatabaseCalciteSchema(String dbName, ReadableCatalog catalog) {
+			this.dbName = dbName;
+			this.catalog = catalog;
+		}
+
+		@Override
+		public Table getTable(String tableName) {
+			LOGGER.info("Getting table '{}' from catalog '{}'", tableName, catalogName);
+
+			ObjectPath tablePath = new ObjectPath(dbName, tableName);
+
+			try {
+				CatalogBaseTable table = catalog.getTable(tablePath);
+
+				LOGGER.info("Successfully got table '{}' from catalog '{}'", tableName, catalogName);
+
+				// TODO: [FLINK-12257] Convert CatalogBaseTable to org.apache.calcite.schema.Table
+				//  so that planner can use unified catalog APIs
+				return null;
+			} catch (TableNotExistException e) {
+				throw new CatalogException(e);
+			}
+		}
+
+		@Override
+		public Set<String> getTableNames() {
+			try {
+				return new HashSet<>(catalog.listTables(dbName));
+			} catch (DatabaseNotExistException e) {
+				throw new CatalogException(e);
+			}
+		}
+
+		@Override
+		public RelProtoDataType getType(String name) {
+			return new RelProtoDataType() {
+				@Override
+				public RelDataType apply(RelDataTypeFactory relDataTypeFactory) {
+					return relDataTypeFactory.createSqlType(SqlTypeName.valueOf(name));
+				}
+			};
+		}
+
+		@Override
+		public Set<String> getTypeNames() {
+			return new HashSet<>();
+		}
+
+		@Override
+		public Collection<Function> getFunctions(String s) {
+			return new HashSet<>();
+		}
+
+		@Override
+		public Set<String> getFunctionNames() {
+			return new HashSet<>();
+		}
+
+		@Override
+		public Schema getSubSchema(String s) {
+			throw new UnsupportedOperationException(
+				"DatabaseCalciteSchema does not support getSubSchema()");
+		}
+
+		@Override
+		public Set<String> getSubSchemaNames() {
+			throw new UnsupportedOperationException(
```

Comment:
Why not empty set?

Reply:
Similar to above, does it make sense to allow trying to get a database out of the table layer?